python/cudf_polars/cudf_polars/typing/__init__.py

```diff
@@ -147,28 +157,114 @@ def set_udf(
 
 U_contra = TypeVar("U_contra", bound=Hashable, contravariant=True)
 V_co = TypeVar("V_co", covariant=True)
+StateT_co = TypeVar("StateT_co", covariant=True)
 NodeT = TypeVar("NodeT", bound="nodebase.Node[Any]")
+StateT = TypeVar("StateT")
 
 
-class GenericTransformer(Protocol[U_contra, V_co]):
+class GenericTransformer(Protocol[U_contra, V_co, StateT_co]):
     """Abstract protocol for recursive visitors."""
 
     def __call__(self, __value: U_contra) -> V_co:
         """Apply the visitor to the node."""
         ...
 
     @property
-    def state(self) -> Mapping[str, Any]:
+    def state(self) -> StateT_co:
         """Arbitrary immutable state."""
         ...
 
 
+class ASTState(TypedDict):
+    """
+    State for AST transformation in :mod:`cudf_polars.dsl.to_ast`.
+
+    Parameters
+    ----------
+    for_parquet
+        Indicator for whether this transformation should provide an expression
+        suitable for use in parquet filters.
+
+        If ``for_parquet`` is ``False``, the dictionary should contain a
+    """
+
+    for_parquet: bool
+
+
+class GenericState(Generic[NodeT], TypedDict):
+    """
+    State for :func:`cudf_polars.dsl.utils.replace.replace`.
+
+    Parameters
+    ----------
+    replacements
+        Mapping from nodes to be replaced to their replacements.
+        ``GenericState`` is generic over the type of these nodes.
+    """
+
+    replacements: Mapping[NodeT, NodeT]
+
+
+class ExprExprState(TypedDict):
+    """
+    State used for AST transformation in :mod:`cudf_polars.dsl.to_ast`.
+
+    Parameters
+    ----------
+    name_to_index
+        Mapping from column names to column indices in the table
+        eventually used for evaluation.
+    table_ref
+        pylibcudf `TableReference` indicating whether column
+        references are coming from the left or right table.
+    """
+
+    name_to_index: Mapping[str, int]
+    table_ref: plc.expressions.TableReference
+
+
+class ExprDecomposerState(TypedDict):
+    """
+    State for :func:`cudf_polars.experimental.expressions.decompose_expr_graph`.
+
+    Parameters
+    ----------
+    input_ir
+        IR of the input expression.
+    input_partition_info
+        Partition info of the input expression.
+    """
+
+    input_ir: ir.IR
+    input_partition_info: PartitionInfo
+    config_options: ConfigOptions
+    unique_names: Generator[str, None, None]
+
+
+class LowerIRState(TypedDict):
+    """
+    State used for transformations in :func:`cudf_polars.experimental.parallel.lower_ir_node`.
+
+    Parameters
+    ----------
+    config_options
+        GPUEngine configuration options.
+    """
+
+    config_options: ConfigOptions
```

Comment:
I think we should now move these closer to point of use (certainly the to_ast one!)

Reply:
Agreed, this is done in 73fae84222.